Support: Update presales and precancellation toggles to reflect API

jeremeylduvall opened the pull request:
As part of the work being done in p7QphF-ai-p2, this PR is meant to update the precancellation and presales chat display in Calypso to reflect the new toggles made available in D8590-code. You can find a link to the current toggles in that diff as well. It's still in progress, and I'll update this PR as I finish it up.

The main things this updates:
- I added a selector to `client/state/ui/olark/selectors.js` that indicates whether precancellation and/or presales chat is turned on.
- We're removing the existing A/B tests for `presaleChatButton` and `chatOfferOnCancel`. The latter wasn't being used and the former has been active since 3/28/17.
- When a user is either cancelling or removing a Business or Premium plan from their site and we have precancellation chat turned on, we now show a Happychat step during the cancellation survey.
- When a user adds a Business plan to their cart and we have presales chat turned on, they'll see a "Need help? Chat with us" button on the cart.

**Todo:**
- [x] Add tests for new selectors, checkout steps, and purchase steps.
- [x] Fix existing tests for CircleCi
- [x] Post detailed testing instructions
- [ ] Update e2e tests with A/B test changes

jeremeylduvall commented:
I rebased this PR with master and updated the various tests that were failing previously. The existing tests should pass now, and both the new Olark state selectors and the new marketing cancellation survey steps have tests as well. I have _not_ yet updated the automated e2e tests with the A/B tests that I removed. I can do that after this is approved and merged.

This PR grew a bit more than I was expecting, but I'm hoping the changes are straightforward and tested well enough to not divide it up. Let me know if you would prefer I split it though. Pinging @markryall @southp @unDemian for some eyes.

The main flows that are affected are the presales and precancellation flows.

### Test presales
1. Visit the livechat toggles available in MC for Happiness. Make sure presales is turned "On." Further instruction is available here: D8590-code.
2. Visit calypso.localhost:3000/plans/ on a site that has a Free, Personal, or Premium plan currently. 
3. Click "Upgrade" to add a Business plan to your cart.

Assuming the presales toggle is turned on and we have operators available, you should see the option to chat on the checkout form as shown here:

<img width="732" alt="screen shot 2018-01-04 at 6 44 50 am" src="https://user-images.githubusercontent.com/7240478/34566177-e008ee60-f11a-11e7-9e6b-aac68108acf8.png">

Try revisiting the presales toggle and turning the option off. Refresh the checkout page. The option to chat should not be available.

### Test precancellation
1. Visit the livechat toggles available in MC for Happiness. Make sure precancellation is turned "On." Further instruction is available here: D8590-code.
2. Visit http://calypso.localhost:3000/me/purchases and select a Business, Premium, or Personal plan.
3. Select anything but "I could not install..." on the first question.
4. Click "Next step" to move beyond the first question.

You should be offered the option to chat with us as shown here:

<img width="568" alt="screen shot 2018-01-04 at 6 54 17 am" src="https://user-images.githubusercontent.com/7240478/34566524-24d1cb24-f11c-11e7-83d3-65122f1fe933.png">

If you turn the toggle off, you should go directly from step 1 to the cancellation/remove step. This should work similarly for cancellations and refunds.

If you would like to run the local tests that I updated, you can do so with these commands:
- Olark selector: `npm run test-client client/state/ui/olark/test/selectors.js`
- Cancellation survey steps: `npm run test-client client/components/marketing-survey/cancel-purchase-form/test/stepsForProductAndSurvey.js`

southp commented on client/components/marketing-survey/cancel-purchase-form/stepsForProductAndSurvey.js:
```diff
@@ -44,6 +39,15 @@ export default function stepsForProductAndSurvey( survey, product, canChat ) {
+			includesProduct( PERSONAL_PREMIUM_PLANS, product ) ) &&
+		precancellation
+	) {
+		return [ steps.INITIAL_STEP, steps.HAPPYCHAT_STEP, steps.FINAL_STEP ];
```
This array is used across multiple places: this line, line 52, and in the test suite as DEFAULT_STEPS_WITH_HAPPYCHAT. I think that makes it a good candidate as a exported module constant. Maybe `steps.js` is a good place? e.g.
```
export const DEFAULT_STEPS_WITH_HAPPYCHAT = [ steps.INITIAL_STEP, steps.HAPPYCHAT_STEP, steps.FINAL_STEP ];
```

That way, we can replace these two lines with it and reuse the same value in the test suite. In the future, if we need to add more steps or make it less, we only need to update one place.

southp commented on client/me/purchases/cancel-purchase/button.jsx:
```diff
@@ -368,6 +374,7 @@ export default connect(
 	state => ( {
 		isChatAvailable: isHappychatAvailable( state ),
 		isChatActive: hasActiveHappychatSession( state ),
+		cancellationChatAvailable: isCancellationChatAvailable( state ),
```
Based on its purpose, it would be clearer if we name it consistently as `precancellationChatAvailable` and `isPrecancellationChatAvailable()`.

southp commented on client/state/ui/olark/selectors.js:
```diff
@@ -51,3 +51,21 @@ export function isChatAvailable( state, context ) {
+ * @param   {Object}  state  Global state tree
+ * @returns {Boolean}        true, when precancellation is available
+ */
+export function isCancellationChatAvailable( state ) {
```
The same idea from the above comment, let's rename "cancellation" as "precancellation" :)

jeremeylduvall replied:
Sounds good! Will update.

southp commented on client/components/marketing-survey/cancel-purchase-form/stepsForProductAndSurvey.js:
```diff
@@ -20,17 +20,12 @@ const JETPACK_PAID_PLANS = [
+	survey,
+	product,
+	canChat,
+	precancellation = false
```
For me, it isn't quite clear why `precancellation` would have default argument as `false` while the others don't. In my opinion, default arguments should only used when it's the most common sensical value. Could we consider to remove this and fill in the call site with proper value?

jeremeylduvall replied:
Just to be clear, I remove the default argument in the function declaration so now we have this:

```js
export default function stepsForProductAndSurvey(
	// ...
	precancellation,
) {
	// ...
}
```

Now, wherever I call the function, I'm providing a default value during the destructuring:

``` js

const { precancellationChatAvailable = false } = this.props;

const steps = stepsForProductAndSurvey(
	// ...
	precancellationChatAvailable,
);
```

Just wanted to make sure that's what you're referring to!

jeremeylduvall commented:
@southp Thanks for the review! I believe I addressed your concerns mentioned above. I left inline comments when necessary, but here are the changes:

- I added `DEFAULT_STEPS_WITH_HAPPYCHAT` as an export to `steps.js` so we can reuse that elsewhere.
- I renamed the cancellation selector and necessary props to `isPrecancellationChatAvailable` and `precancellationChatAvailable`.
- I modified how we're defaulting `precancellationChatAvailable` to false as described in [this example](https://github.com/Automattic/wp-calypso/pull/21200#discussion_r162101424).
- I removed the duplicate chat option by temporarily adding a check for `happychat_step` [here](https://github.com/Automattic/wp-calypso/blob/3ce3034a07e48e13ec1336e752e600a0f0691981/client/me/purchases/remove-purchase/index.jsx#L358)

Everything still functions as it did originally and tests are all passing. Here's a snapshot of the happychat step during removal without the duplicate happychat button:

<img width="548" alt="screen shot 2018-01-17 at 9 47 40 am" src="https://user-images.githubusercontent.com/7240478/35056231-7861ed6c-fb6e-11e7-8a4c-68ad8c4d1e6e.png">

jeremeylduvall commented:
Following the merge of #21488, I moved the selectors out of `state/ui/olark` and dropped them in `state/happychat/selectors` instead. I have a few questions (@jsnajdr perhaps you can help answer after your work on removing Olark):

1. I'm still relying on `state.ui.olark.availability` to expose the live chat options mentioned in D8590-code. That still seems to be available after your changes?
2. I'm a bit stuck on the rebase because I modified `client/state/ui/olark/selectors.js` and `client/state/ui/olark/test/selectors.js` on my branch, but they're removed on master. I tried deleting the files on my branch and then finishing the rebase, but it's still stuck:

> CONFLICT (modify/delete): client/state/ui/olark/selectors.js deleted in HEAD and modified in Help: Add selector for presales and precancellation chat. Version Help: Add selector for presales and precancellation chat of client/state/ui/olark/selectors.js left in tree.

Any advice on how to finish the rebase?

jsnajdr commented:
Hello @jeremeylduvall!

I recommend to store the presale and precancellation availability flags in `state.happychat.availability`. The `state.ui.olark` tree was completely removed and doesn't exist any more.

To finish the rebase, I think you need to `git rm` the conflicting files, then `git add` the directory with the deleted files, and then `git rebase --continue`. I usually look up the rarely used workflows on StackOverflow ;)

jeremeylduvall commented:
Thanks! Merge conflicts resolved. It through me for a loop since the deleted files were actually modified in multiple commits in this branch so I had to go through the process a few times to clean up all the commits. Either way, it's all good now!

I marked this "In Progress" while I work on adding the availability under `state.happychat.availability`. I'll have that done this week.

southp commented:
Thanks for the effort! Please ping us when you are done, and we will help push this through 🙇🏻

jeremeylduvall commented:
I moved the presales and precancellation to the Happychat state tree. Quick recap:

- Now, when we request `requestHappychatEligibility()`, which is run [on boot](https://github.com/Automattic/wp-calypso/blob/0cb28766c3c86da451367ac7ee35e39739863910/client/boot/project/wordpress-com.js#L214) and [on the removal of a purchase](https://github.com/Automattic/wp-calypso/blob/0cb28766c3c86da451367ac7ee35e39739863910/client/state/purchases/actions.js#L59), we also set the [presale/precancellation availability](https://github.com/Automattic/wp-calypso/blob/0661eab3a3b903a44a03a7aa2ec125fed0eaba4f/client/state/happychat/user/actions.js#L31).
- I added the presale/precancellation availability to `state.happychat.user` [here](https://github.com/Automattic/wp-calypso/blob/0661eab3a3b903a44a03a7aa2ec125fed0eaba4f/client/state/happychat/user/reducer.js#L51) ([schema updated](https://github.com/Automattic/wp-calypso/blob/0661eab3a3b903a44a03a7aa2ec125fed0eaba4f/client/state/happychat/user/schema.js#L17)). I'm not sure if this is a good place for it or not, but the other choices didn't seem appropriate either (chat is all about the individual chat, connection is about Happychat overall, and UI seemed to be all about Happychat actions).
- Selectors and tests are updated to reflect the change in the state tree.

Going to flip this back to "Needs Review" @southp. Let me know what you think!

southp commented on client/state/happychat/selectors/is-precancellation-chat-available.js:
```diff
@@ -0,0 +1,8 @@
+ * @returns {Boolean}        true, when precancellation is available
+ */
+export default function isPrecancellationChatAvailable( state ) {
+	return state.happychat.user.isPresalesPrecancellationEligible.precancellation;
```
Since there is no guarantee that the substate `happyisPresalesPrecancellationEligible` isn't null from the reducer side, I'd suggest we use `lodash.get` to make this safer. e.g.

```
import { get} from 'lodash';
...
return get( state, 'happychat.user.isPresalesPrecancellationEligible.presale', false );
```

The same goes to `isPresaleChatAvailable()` below.

That way, we don't have to default the connected props from components, either.

southp commented on client/me/purchases/remove-purchase/index.jsx:
```diff
@@ -117,12 +118,18 @@ class RemovePurchase extends Component {
+			selectedPurchase,
+			isChatAvailable,
+			isChatActive,
+			precancellationChatAvailable = false,
```
Since this is a "connected property" from redux state, I personally prefer to let the selector handling this. See the inlined comment below for more details :)

jeremeylduvall commented:
Thanks! Modified to use lodash `get`. I'll give the CircleCI tests some time to pass then merge later today when I'm online.
